allow users list their own access tokens

stlaz opened the pull request:
Adds an endpoint that mirrors the oauthaccesstokens that only belong to the requesting user. The new endpoint allows these operations:
- get
- list
- watch
- delete

cc @deads2k @sttts 

enhancement ref: https://github.com/openshift/enhancements/pull/507

sttts commented on pkg/oauth/apis/oauth/types.go:
```diff
@@ -224,3 +224,19 @@ type RedirectReference struct {
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PersonalAccessToken struct {
```
my vote goes to UserOauthAccessToken

sttts commented on pkg/oauth/apiserver/apiserver.go:
```diff
@@ -139,11 +140,16 @@ func (c *completedConfig) newV1RESTStorage() (map[string]rest.Storage, error) {
 	v1Storage["oAuthAccessTokens"] = accessTokenStorage
 	v1Storage["oAuthClients"] = clientStorage
 	v1Storage["oAuthClientAuthorizations"] = clientAuthorizationStorage
+	v1Storage["personalTokens"] = personalTokensDelegate
```
needs some naming unification following the enhancement.

sttts commented on pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func (r *REST) List(ctx context.Context, options *metainternal.ListOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
```
can it be empty?

sttts replied:
nvmd. It's your func below.

stlaz replied:
`getUserFromContext` checks for an empty username and errors out in that case

sttts commented on pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go:
```diff
@@ -0,0 +1,186 @@
+	// the current field selector cannot make any assumptions over the "userName" field
+	for _, req := range newOpts.FieldSelector.Requirements() {
+		if req.Field == "userName" {
+			return nil, errors.NewUnauthorized("the userName field selector is not allowed")
```
can we alternatively just add our userName constraint? If the user adds another one for a different user, the set will just be empty. That's a cleaner and maybe safer solution.

sttts replied:
You actually do that below. I.e. we could just drop this whole check?

sttts commented on pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go:
```diff
@@ -0,0 +1,186 @@
+		return nil, err
+	}
+
+	tokenList := tokenListGenericObj.(*oauthapi.OAuthAccessTokenList)
```
```
tokenList, ok := ...
if !ok {
```

The current code panics.

sttts commented on pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go:
```diff
@@ -0,0 +1,186 @@
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	if _, err := r.Get(ctx, name, &metav1.GetOptions{}); err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
```
here is a small risk for a race: you have to pass the uid of the returned object in the options below, and probably also do some translation of GetOptions to DeleteOptions above.

sttts commented on pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go:
```diff
@@ -0,0 +1,186 @@
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		return "", kerrors.NewForbidden(oauth.Resource("personaltokens"), "", fmt.Errorf("unable to get personal tokens without a user in context"))
```
Or just 404.

sttts commented on pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go:
```diff
@@ -0,0 +1,186 @@
+func oauthAccessTokenListToUserOAuthAccessTokenList(l *oauthapi.OAuthAccessTokenList) *oauthapi.PersonalAccessTokenList {
+	ret := oauthapi.PersonalAccessTokenList{}
+	for _, t := range l.Items {
+		ret.Items = append(ret.Items, oauthapi.PersonalAccessToken{OAuthAccessToken: t})
```
do we use this pattern of embedding elsewhere? If not, I would be slightly concerned that things might break if TypeMeta and ObjectMeta are not at the toplevel of the type.

stlaz replied:
I assume it's novel, I wasn't looking for it elsewhere

sttts replied:
we will see. If you have it working, maybe everything is fine.

sttts commented on pkg/oauth/printers/internalversion/printer.go:
```diff
@@ -159,6 +165,27 @@ func printOAuthAccessToken(oauthAccessToken *oauthapi.OAuthAccessToken, options
+		created,
+		expires,
+		personalAccessToken.RedirectURI,
+		strings.Join(personalAccessToken.Scopes, ","),
```
@Anandnatraj these are the columns with oc.

stlaz commented:
/retest
I wonder why it's a baremetal operator that failed on e2e-aws

stlaz commented:
Addressed all comments. The PR now depends on https://github.com/openshift/api/pull/774 (aside from the enhancement merging)

deads2k commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,216 @@
+	tableConvertor rest.TableConvertor
+}
+
+var _ rest.Lister = &REST{}
```
isn't there a standard REST you can match against?

deads2k replied:
> isn't there a standard REST you can match against?

ah, no create or update.  Add comment.

deads2k commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,216 @@
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken GET result to its typed version"))
+	}
+
+	if token.UserName != ctxUserName || !isTokenSha256Token(token.Name) { // don't reveal other people tokens' hashes or non-hashed tokens
```
make this a filter func so you can re-use it everywhere.

deads2k commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,216 @@
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
+	}
+
+	deletedTyped := deletedObj.(*oauthapi.UserOAuthAccessToken)
```
check user

stlaz replied:
It's checked in the `Get()` method but I agree that explicit is better than implicit and who knows what can happen to the `Get()` in the future. I added that check.

deads2k commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,216 @@
+func oauthAccessTokenListToUserOAuthAccessTokenList(l *oauthapi.OAuthAccessTokenList) *oauthapi.UserOAuthAccessTokenList {
+	ret := oauthapi.UserOAuthAccessTokenList{}
+	for _, t := range l.Items {
+		if !isTokenSha256Token(t.Name) {
```
confirm user just in case

deads2k commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go:
```diff
@@ -0,0 +1,106 @@
+					continue
+				}
+
+				if !isTokenSha256Token(tokenOrig.Name) {
```
filter by name too.

Also, I think there's a first class filter concept.

stlaz replied:
Added the filtering.

Not sure whether there's a better, first class filtering. The channel I'm "borrowing" belongs to another watcher that seems to only allow conventional ListOptions

sttts commented:
/retest

stlaz commented:
/hold cancel
this is not a prototype anymore

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	newOpts.FieldSelector = fields.AndSelectors(
+		fields.Set{"userName": userName}.AsSelector(),
```
we need an e2e test for this

stlaz replied:
I added some e2e tests here: https://github.com/openshift/origin/pull/25697

stlaz replied:
moved e2e tests to this repo

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,228 @@
+	// it then continues to watch from the last known RV
+	// This prevents the initial objects to be returned twice to the client
+	// (objects from the list + from the next watch with an uninitialized from-RV)
+	ret.ResourceVersion = l.ResourceVersion
```
why not everything else from the meta fields?

stlaz replied:
I thought I would limit the amount of information to only that we really need. I however understand that it is hard to objectively tell which information is useful and necessary and which is not, I think I'll add all of the metadata, although perhaps conditionally when at least _some_ items appear in the filtered slice?

stlaz replied:
actually, never mind, I think we can always set it

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,228 @@
+
+	deletedObj, err := r.Get(ctx, name, &metav1.GetOptions{TypeMeta: options.TypeMeta, ResourceVersion: deletedRV})
+	if err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
```
why not nil as object?

stlaz replied:
I was not sure how exactly this should be handled and the docs of the `GracefulDeleter` made me think this was the right way, but reading the code of the generic `Store`, it seems that `nil` would be the right option.

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,228 @@
+
+	case newOpts.Preconditions != nil && newOpts.Preconditions.UID != nil && *newOpts.Preconditions.UID != deletedTyped.UID:
+		// the UID seeked to be deleted no longer exists/never existed
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
```
wouldn't we get this for free from the delete call below?

stlaz replied:
I think so... I wanted to prevent user-supplied options being passed to the actual privileged call, but given we make sure the preconditions/name combination does not exist, we should be able to get to the same result.

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,228 @@
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		klog.V(4).Infof("no user found in the context")
```
do we want this? When can it happen?

stlaz replied:
I am not sure whether it's at all possible, it felt like a safe precaution.

sttts replied:
Would remove the log output here.

sttts replied:
or even turn this into `string, bool, error` and return the NotFound in the call sites.

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go:
```diff
@@ -0,0 +1,112 @@
+
+// OAuthAccessTokenWatcher implements a Watcher interface to watch OAuthAccessTokens
+// and filter them to match only those for the current user
+type OAuthAccessTokenWatcher struct {
```
is this taken from anywhere?

stlaz replied:
I was looking for inspiration in the watchers of k8s.io/apimachinery/pkg/watch but I ended up mixing some of the basic bits of the openshift-apiserver's project watcher with custom resource proxying logic

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go:
```diff
@@ -0,0 +1,112 @@
+	defer w.mutex.Unlock()
+	if !w.stopped {
+		w.stopped = true
+		// forward the Stop() request
```
no additional value of the comment

stlaz commented:
The latest revision adds fixes that:
- prevent panic on fieldselector being nil
- added `clientName` field selector for the new API type
- fixed an issue preventing Get from working correctly

Todo:
- move e2e tests here

stlaz commented:
/hold
now depends on https://github.com/openshift/api/pull/788 and https://github.com/openshift/client-go/pull/167

sttts commented:
/retest

stlaz commented:
/hold cancel
removed fake bumps as dependencies merged
https://github.com/openshift/release/pull/13807 adds the ci configuration for the component e2e tests

stlaz commented:
/retest

stlaz commented:
rbac for the new endpoints added by https://github.com/openshift/cluster-authentication-operator/pull/381

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+	"github.com/openshift/oauth-apiserver/pkg/printerstorage"
+)
+
+// rest implements a RESTStorage for access tokens against etcd
```
nit: for access tokens that the current user owns (selected by label)

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+
+	userName := userInfo.GetName()
+	if len(userName) == 0 {
+		klog.V(4).Info("user in context has zero-length username")
```
same here. No username means nothing to see. `return "", false, nil` and NotFound where this func is called.

stlaz replied:
what's the use of error, then?

sttts replied:
right, just get rid of it. I prefer to put logic for errors returned to the user at the top-level, not deep into helpers.

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+	return &ret
+}
+
+func userFieldSelectorToListOptions(opts *metainternal.ListOptions, userName string) (*metainternal.ListOptions, error) {
```
ListOptionsWithUserNameFilter

p0lyn0mial replied:
this method doesn't return `error` - drop it.

sttts commented on test/e2e/useroauthaccesstokens_test.go:
```diff
@@ -0,0 +1,560 @@
+			name:            "set label selector to get own tokens and of others",
+			userToken:       mirkaTokenString,
+			userName:        "mirka",
+			labelSelector:   parseLabelSelectorOrDie("randomLabel notin (mirka,franta)"),
```
would add the simple case too: userName == somebody else.

stlaz replied:
there's a fieldSelector like this, but I'll add a label to one of the basic tokens above and will attempt to list it

sttts commented on test/e2e/useroauthaccesstokens_test.go:
```diff
@@ -0,0 +1,560 @@
+			name:         "delete an own token",
+			userToken:    frantaTokenString,
+			userName:     "mirka",
+			getTokenName: "sha256~" + testNS + "franta0",
```
also test with a wrong UID.

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+
+	tokenListGenericObj, err := r.accessTokenStorage.List(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
```
should we mask the err ?

stlaz replied:
I think we *should* be fine with the error as is

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+
+	tokenList, ok := tokenListGenericObj.(*oauthapi.OAuthAccessTokenList)
+	if !ok {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken LIST result to its typed version"))
```
could we be more specific and log the underlying type for `tokenListGenericObj` and `OAuthAccessTokenList` ?

stlaz replied:
This error gets propagated to the user, does it not? I think it might be confusing when they see a mention of a different type from what they requested

p0lyn0mial replied:
When the conversion can fail? Can it only happen when the wiring is wrong? 

In that case, I would prefer to see a very clear and specific error so that I can debug it.

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+		newOpts = opts.DeepCopy()
+	}
+
+	userSelector := fields.Set{"userName": userName}.AsSelector()
```
what if `opts` already specifies the field selector for the `userName`?

I think we should be fine since we have an additional check `isValidUserToken` but maybe we could forbid specifying that particular selector?

sttts replied:
we call `AndSelectors` which will make sure our user is always checked for. If the user checks for another one, the selector will always be false. That's what we want. Compare e2e tests. There is one for that case.

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go:
```diff
@@ -0,0 +1,111 @@
+				event.Object = (*oauthapi.UserOAuthAccessToken)(tokenOrig)
+				w.outgoing <- event
+			}
+		case <-ctx.Done(): // user cancel
```
in that case, we won't set `w.stopped = true` and won't close the  `close(w.stopCh)`

stlaz replied:
good catch

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// StopChan returns stop channel
+func (w *OAuthAccessTokenWatcher) StopChan() <-chan struct{} {
```
is this method used ?

stlaz replied:
I copied the interface of the `apimachinery/pkg/watch.ProxyWatcher` struct, but it looks like this and `Stopping` may indeed be safe to remove

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// Stopping returns true if Stop() has been called
+func (w *OAuthAccessTokenWatcher) Stopping() bool {
```
is this method used?

p0lyn0mial commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -0,0 +1,218 @@
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken LIST result to its typed version"))
+	}
+
+	return oauthAccessTokenListToUserOAuthAccessTokenList(tokenList, ctxUserName), nil
```
at this point `tokenList != nil`, right ?

stlaz replied:
yes, that should hold

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -57,15 +55,12 @@ func (r *REST) NewList() runtime.Object {
-		return nil, err
+	ctxUserName, ok := getUserFromContext(ctx)
+	if !ok {
+		return nil, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
```
Shouldn't this not rather be just an empty result?

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -81,18 +76,18 @@ func (r *REST) List(ctx context.Context, options *metainternal.ListOptions) (run
-		return nil, err
+	ctxUserName, ok := getUserFromContext(ctx)
+	if !ok {
+		return nil, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
```
as for List, but with second argument to be name

sttts commented on pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go:
```diff
@@ -133,9 +125,9 @@ func (r *REST) NamespaceScoped() bool {
-		return nil, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	ctxUserName, ok := getUserFromContext(ctx)
+	if !ok {
+		return nil, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
```
keep the name

stlaz commented:
/retest

sttts commented:
/lgtm
/approve

stlaz commented:
The component tests are currently failing because https://github.com/openshift/cluster-authentication-operator/pull/381 hasn't merged yet

stlaz commented:
/hold
wait for rbac deps to merge

sttts commented:
/retest

stlaz commented:
we still need https://github.com/openshift/cluster-authentication-operator/pull/381 for the tests to pass :/

stlaz commented:
/hold cancel
the operator PR merged. We'll probably need a few retests before the RBAC is actually there

stlaz commented:
/retest
